include/revng/ABI/CallingConventionTrait.h

```diff
@@ -0,0 +1,1083 @@
+  ///
+  /// \note: If `MaximumGPRsPerScalarArgument` is equal to 0, it means that
+  /// the ABI does not allow scalar arguments to use GPRs.
+  ///
+  /// \note: If an argument doesn't fit into the specified registers or
+  /// uses irregular padding, the registers are not used and the object is
+  /// passed using the memory (stack, pointer-to-copy, etc.).
+  static constexpr size_t MaximumGPRsPerScalarArgument = 0;
+
+  /// States the maximum number of GPRs available to return a single scalar
+  /// (`int`, `__int128`, pointer, etc.) value, meaning that it can only be
+  /// returned in the GPRs if `MaximumGPRsPerScalarReturnValue` is less than
+  /// or equal to the number of the registers required to fit the object
+  /// including padding.
+  ///
+  /// \note: If `MaximumGPRsPerScalarReturnValue` is equal to 0, it means
+  /// that the ABI does not allow scalar return values to use GPRs.
+  ///
+  /// \note: If a return value doesn't fit into the specified registers or
+  /// uses irregular padding, the registers are not used and the object is
+  /// passed using the memory (stack, pointer-to-copy, etc.).
+  static constexpr size_t MaximumGPRsPerScalarReturnValue = 0;
+
+  /// States the required size for a single stack argument in bytes.
+  static constexpr size_t MinimumStackArgumentSize = 1;
+
+  /// Stores the list of general purpose registers allowed to be used for
+  /// passing arguments and the order they are to be used in.
+  static constexpr std::array<model::Register::Values, 0>
+    GeneralPurposeArgumentRegisters = {};
+
+  /// Stores the list of general purpose registers allowed to be used for
+  /// returning values and the order they are to be used in.
+  static constexpr std::array<model::Register::Values, 0>
+    GeneralPurposeReturnValueRegisters = {};
+
+  /// Stores the list of vector registers allowed to be used for passing
+  /// arguments and the order they are to be used in.
+  static constexpr std::array<model::Register::Values, 0>
+    VectorArgumentRegisters = {};
+
+  /// Stores the list of vector registers allowed to be used for returning
+  /// values and the order they are to be used in.
+  static constexpr std::array<model::Register::Values, 0>
+    VectorReturnValueRegisters = {};
+
+  /// Stores the list of registers for which the ABI requires the callee to
+  /// preserve the value, meaning that when the callee returns, the value of
+  /// those registers must be the same as it was when the function was called.
+  static constexpr std::array<model::Register::Values, 0>
+    CalleeSavedRegisters = {};
+
+  /// Specifies a register to be used for returning (or even passing,
+  /// depending on ABI) the pointer to the memory used for returning
+  /// copies of big aggregate objects.
+  ///
+  /// Can be `model::Register::Invalid` for ABIs that do not support returning
+  /// values by 'pointer-to-copy'.
+  static constexpr auto ReturnValueLocationRegister = model::Register::Invalid;
```

Comment:
This is for the extra "shadow" argument containing a pointer for the big return aggregate, is that right?
Please in the documentation say something like "shadow" argument, since otherwise sounds like the information is flowing only from the callee to the caller, while the caller is actually using this register to tell the callee where to put the return value.

Reply:
Not always, sometimes it does indeed serve indicate the location of the "argument"-like return value. But some ABIs (64-bit arm, for example) have a dedicated register for that - it doesn't interfere with other arguments.

Reply:
It does not interfere with other arguments but it's still an argument, in a way. Don't you agree?

Reply:
Ok, yes, you're right. What do you think of something like this?
```
/// Specifies a register to be used to pass the pointer to the memory used for returning copies of big aggregate objects on for the ABIs that support such a behaviour. It should be set to `model::Register::Invalid` for the ABIs that do not.
```

Reply:
I see you don't like to call it an argument `:D` but OK, works.

Reply:
Well, since I used the word `pass` it's already pretty much an argument `:D`